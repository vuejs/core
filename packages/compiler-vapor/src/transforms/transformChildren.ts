import { type ElementNode, ElementTypes, NodeTypes } from '@vue/compiler-dom'
import {
  type NodeTransform,
  type TransformContext,
  transformNode,
} from '../transform'
import {
  DynamicFlag,
  type IRDynamicInfo,
  IRNodeTypes,
  isBlockOperation,
} from '../ir'

export const transformChildren: NodeTransform = (node, context) => {
  const isFragment =
    node.type === NodeTypes.ROOT ||
    (node.type === NodeTypes.ELEMENT &&
      (node.tagType === ElementTypes.TEMPLATE ||
        node.tagType === ElementTypes.COMPONENT))

  if (!isFragment && node.type !== NodeTypes.ELEMENT) return

  for (const [i, child] of node.children.entries()) {
    const childContext = context.create(child, i)
    transformNode(childContext)

    const childDynamic = childContext.dynamic

    if (isFragment) {
      childContext.reference()
      childContext.registerTemplate()

      if (
        !(childDynamic.flags & DynamicFlag.NON_TEMPLATE) ||
        childDynamic.flags & DynamicFlag.INSERT
      ) {
        context.block.returns.push(childContext.dynamic.id!)
      }
    } else {
      context.childrenTemplate.push(childContext.template)
    }

    if (
      childDynamic.hasDynamicChild ||
      childDynamic.id !== undefined ||
      childDynamic.flags & DynamicFlag.NON_TEMPLATE ||
      childDynamic.flags & DynamicFlag.INSERT
    ) {
      context.dynamic.hasDynamicChild = true
    }

    context.dynamic.children[i] = childDynamic
  }

  if (!isFragment) {
    processDynamicChildren(context as TransformContext<ElementNode>)
  }
}

function processDynamicChildren(context: TransformContext<ElementNode>) {
  let prevDynamics: IRDynamicInfo[] = []
  let hasStaticTemplate = false
  const children = context.dynamic.children

  for (const [index, child] of children.entries()) {
    if (child.flags & DynamicFlag.INSERT) {
      prevDynamics.push(child)
    }

    if (!(child.flags & DynamicFlag.NON_TEMPLATE)) {
      if (prevDynamics.length) {
        if (hasStaticTemplate) {
          context.childrenTemplate[index - prevDynamics.length] = `<!>`
          prevDynamics[0].flags -= DynamicFlag.NON_TEMPLATE
          const anchor = (prevDynamics[0].anchor = context.increaseId())
          registerInsertion(prevDynamics, context, anchor)
        } else {
          registerInsertion(prevDynamics, context, -1 /* prepend */)
        }
        prevDynamics = []
      }
      hasStaticTemplate = true
    }
  }

  if (prevDynamics.length) {
    registerInsertion(prevDynamics, context)
  }
}

function registerInsertion(
  dynamics: IRDynamicInfo[],
  context: TransformContext,
  anchor?: number,
) {
  for (const child of dynamics) {
    if (child.template != null) {
      // template node due to invalid nesting - generate actual insertion
      context.registerOperation({
        type: IRNodeTypes.INSERT_NODE,
        elements: dynamics.map(child => child.id!),
        parent: context.reference(),
        anchor,
      })
    } else if (child.operation && isBlockOperation(child.operation)) {
      // block types
      child.operation.parent = context.reference()
      child.operation.anchor = anchor
    }
  }
}
