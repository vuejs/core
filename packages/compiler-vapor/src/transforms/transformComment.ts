import {
  type CommentNode,
  type ElementNode,
  NodeTypes,
  type TemplateChildNode,
} from '@vue/compiler-dom'
import type { NodeTransform, TransformContext } from '../transform'
import { DynamicFlag } from '../ir'

export const transformComment: NodeTransform = (node, context) => {
  if (node.type !== NodeTypes.COMMENT) return

  if (getSiblingIf(context as TransformContext<CommentNode>)) {
    context.comment.push(node)
    context.dynamic.flags |= DynamicFlag.NON_TEMPLATE
  } else {
    context.template += `<!--${node.content}-->`
  }
}

export function getSiblingIf(
  context: TransformContext<TemplateChildNode>,
  reverse?: boolean,
): ElementNode | undefined {
  const parent = context.parent
  if (!parent) return

  const siblings = parent.node.children
  let sibling: TemplateChildNode | undefined
  let i = siblings.indexOf(context.node)
  while (reverse ? --i >= 0 : ++i < siblings.length) {
    sibling = siblings[i]
    if (!isCommentLike(sibling)) {
      break
    }
  }

  if (
    sibling &&
    sibling.type === NodeTypes.ELEMENT &&
    sibling.props.some(
      ({ type, name }) =>
        type === NodeTypes.DIRECTIVE &&
        ['else-if', reverse ? 'if' : 'else'].includes(name),
    )
  ) {
    return sibling
  }
}

function isCommentLike(node: TemplateChildNode) {
  return (
    node.type === NodeTypes.COMMENT ||
    (node.type === NodeTypes.TEXT && !node.content.trim().length)
  )
}
