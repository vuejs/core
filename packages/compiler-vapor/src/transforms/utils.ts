import {
  type AttributeNode,
  type DirectiveNode,
  type ElementNode,
  ElementTypes,
  NodeTypes,
  type SimpleExpressionNode,
  type TemplateChildNode,
  type TemplateNode,
  createSimpleExpression,
} from '@vue/compiler-dom'
import { extend } from '@vue/shared'
import {
  type BlockIRNode,
  DynamicFlag,
  type IRDynamicInfo,
  IRNodeTypes,
} from '../ir'

export const newDynamic = (): IRDynamicInfo => ({
  flags: DynamicFlag.REFERENCED,
  children: [],
})

export const newBlock = (node: BlockIRNode['node']): BlockIRNode => ({
  type: IRNodeTypes.BLOCK,
  node,
  dynamic: newDynamic(),
  effect: [],
  operation: [],
  returns: [],
  tempId: 0,
})

export function wrapTemplate(node: ElementNode, dirs: string[]): TemplateNode {
  if (node.tagType === ElementTypes.TEMPLATE) {
    return node
  }

  const reserved: Array<AttributeNode | DirectiveNode> = []
  const pass: Array<AttributeNode | DirectiveNode> = []
  node.props.forEach(prop => {
    if (prop.type === NodeTypes.DIRECTIVE && dirs.includes(prop.name)) {
      reserved.push(prop)
    } else {
      pass.push(prop)
    }
  })

  return extend({}, node, {
    type: NodeTypes.ELEMENT,
    tag: 'template',
    props: reserved,
    tagType: ElementTypes.TEMPLATE,
    children: [extend({}, node, { props: pass } as TemplateChildNode)],
  } as Partial<TemplateNode>)
}

export const EMPTY_EXPRESSION: SimpleExpressionNode = createSimpleExpression(
  '',
  true,
)
