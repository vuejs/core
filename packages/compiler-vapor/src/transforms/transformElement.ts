import { isValidHTMLNesting } from '@vue/compiler-dom'
import {
  type AttributeNode,
  type ComponentNode,
  type ElementNode,
  ElementTypes,
  ErrorCodes,
  NodeTypes,
  type PlainElementNode,
  type SimpleExpressionNode,
  createCompilerError,
  createSimpleExpression,
  isStaticArgOf,
} from '@vue/compiler-dom'
import {
  camelize,
  capitalize,
  extend,
  isBuiltInDirective,
  isVoidTag,
  makeMap,
} from '@vue/shared'
import type {
  DirectiveTransformResult,
  NodeTransform,
  TransformContext,
} from '../transform'
import {
  DynamicFlag,
  IRDynamicPropsKind,
  IRNodeTypes,
  type IRProp,
  type IRProps,
  type IRPropsDynamicAttribute,
  type IRPropsStatic,
  type VaporDirectiveNode,
} from '../ir'
import { EMPTY_EXPRESSION } from './utils'
import { findProp } from '../utils'

export const isReservedProp: (key: string) => boolean = /*#__PURE__*/ makeMap(
  // the leading comma is intentional so empty string "" is also included
  ',key,ref,ref_for,ref_key,',
)

export const transformElement: NodeTransform = (node, context) => {
  return function postTransformElement() {
    ;({ node } = context)
    if (
      !(
        node.type === NodeTypes.ELEMENT &&
        (node.tagType === ElementTypes.ELEMENT ||
          node.tagType === ElementTypes.COMPONENT)
      )
    )
      return

    const isComponent = node.tagType === ElementTypes.COMPONENT
    const isDynamicComponent = isComponentTag(node.tag)
    const propsResult = buildProps(
      node,
      context as TransformContext<ElementNode>,
      isComponent,
      isDynamicComponent,
    )

    let { parent } = context
    while (
      parent &&
      parent.parent &&
      parent.node.type === NodeTypes.ELEMENT &&
      parent.node.tagType === ElementTypes.TEMPLATE
    ) {
      parent = parent.parent
    }
    const singleRoot =
      context.root === parent &&
      parent.node.children.filter(child => child.type !== NodeTypes.COMMENT)
        .length === 1

    ;(isComponent ? transformComponentElement : transformNativeElement)(
      node as any,
      propsResult,
      singleRoot,
      context as TransformContext<ElementNode>,
      isDynamicComponent,
    )
  }
}

function transformComponentElement(
  node: ComponentNode,
  propsResult: PropsResult,
  singleRoot: boolean,
  context: TransformContext,
  isDynamicComponent: boolean,
) {
  const dynamicComponent = isDynamicComponent
    ? resolveDynamicComponent(node)
    : undefined

  let { tag } = node
  let asset = true

  if (!dynamicComponent) {
    const fromSetup = resolveSetupReference(tag, context)
    if (fromSetup) {
      tag = fromSetup
      asset = false
    }

    const dotIndex = tag.indexOf('.')
    if (dotIndex > 0) {
      const ns = resolveSetupReference(tag.slice(0, dotIndex), context)
      if (ns) {
        tag = ns + tag.slice(dotIndex)
        asset = false
      }
    }

    if (asset) {
      // self referencing component (inferred from filename)
      if (context.selfName && capitalize(camelize(tag)) === context.selfName) {
        // generators/block.ts has special check for __self postfix when generating
        // component imports, which will pass additional `maybeSelfReference` flag
        // to `resolveComponent`.
        tag += `__self`
      }
      context.component.add(tag)
    }
  }

  context.dynamic.flags |= DynamicFlag.NON_TEMPLATE | DynamicFlag.INSERT
  context.dynamic.operation = {
    type: IRNodeTypes.CREATE_COMPONENT_NODE,
    id: context.reference(),
    tag,
    props: propsResult[0] ? propsResult[1] : [propsResult[1]],
    asset,
    root: singleRoot,
    slots: [...context.slots],
    once: context.inVOnce,
    dynamic: dynamicComponent,
  }
  context.slots = []
}

function resolveDynamicComponent(node: ComponentNode) {
  const isProp = findProp(node, 'is', false, true /* allow empty */)
  if (!isProp) return

  if (isProp.type === NodeTypes.ATTRIBUTE) {
    return isProp.value && createSimpleExpression(isProp.value.content, true)
  } else {
    return (
      isProp.exp ||
      // #10469 handle :is shorthand
      extend(createSimpleExpression(`is`, false, isProp.arg!.loc), {
        ast: null,
      })
    )
  }
}

function resolveSetupReference(name: string, context: TransformContext) {
  const bindings = context.options.bindingMetadata
  if (!bindings || bindings.__isScriptSetup === false) {
    return
  }

  const camelName = camelize(name)
  const PascalName = capitalize(camelName)
  return bindings[name]
    ? name
    : bindings[camelName]
      ? camelName
      : bindings[PascalName]
        ? PascalName
        : undefined
}

function transformNativeElement(
  node: PlainElementNode,
  propsResult: PropsResult,
  singleRoot: boolean,
  context: TransformContext<ElementNode>,
) {
  const { tag } = node
  const { scopeId } = context.options

  let template = ''

  template += `<${tag}`
  if (scopeId) template += ` ${scopeId}`

  const dynamicProps: string[] = []
  if (propsResult[0] /* dynamic props */) {
    const [, dynamicArgs, expressions] = propsResult
    context.registerEffect(expressions, {
      type: IRNodeTypes.SET_DYNAMIC_PROPS,
      element: context.reference(),
      props: dynamicArgs,
      root: singleRoot,
    })
  } else {
    for (const prop of propsResult[1]) {
      const { key, values } = prop
      if (key.isStatic && values.length === 1 && values[0].isStatic) {
        template += ` ${key.content}`
        if (values[0].content) template += `="${values[0].content}"`
      } else {
        dynamicProps.push(key.content)
        context.registerEffect(values, {
          type: IRNodeTypes.SET_PROP,
          element: context.reference(),
          prop,
          root: singleRoot,
          tag,
        })
      }
    }
  }

  template += `>` + context.childrenTemplate.join('')
  // TODO remove unnecessary close tag, e.g. if it's the last element of the template
  if (!isVoidTag(tag)) {
    template += `</${tag}>`
  }

  if (singleRoot) {
    context.ir.rootTemplateIndex = context.ir.template.length
  }

  if (
    context.parent &&
    context.parent.node.type === NodeTypes.ELEMENT &&
    !isValidHTMLNesting(context.parent.node.tag, tag)
  ) {
    context.reference()
    context.dynamic.template = context.pushTemplate(template)
    context.dynamic.flags |= DynamicFlag.INSERT | DynamicFlag.NON_TEMPLATE
  } else {
    context.template += template
  }
}

export type PropsResult =
  | [dynamic: true, props: IRProps[], expressions: SimpleExpressionNode[]]
  | [dynamic: false, props: IRPropsStatic]

export function buildProps(
  node: ElementNode,
  context: TransformContext<ElementNode>,
  isComponent: boolean,
  isDynamicComponent?: boolean,
): PropsResult {
  const props = node.props as (VaporDirectiveNode | AttributeNode)[]
  if (props.length === 0) return [false, []]

  const dynamicArgs: IRProps[] = []
  const dynamicExpr: SimpleExpressionNode[] = []
  let results: DirectiveTransformResult[] = []

  function pushMergeArg() {
    if (results.length) {
      dynamicArgs.push(dedupeProperties(results))
      results = []
    }
  }

  for (const prop of props) {
    if (prop.type === NodeTypes.DIRECTIVE && !prop.arg) {
      if (prop.name === 'bind') {
        // v-bind="obj"
        if (prop.exp) {
          dynamicExpr.push(prop.exp)
          pushMergeArg()
          dynamicArgs.push({
            kind: IRDynamicPropsKind.EXPRESSION,
            value: prop.exp,
          })
        } else {
          context.options.onError(
            createCompilerError(ErrorCodes.X_V_BIND_NO_EXPRESSION, prop.loc),
          )
        }
        continue
      } else if (prop.name === 'on') {
        // v-on="obj"
        if (prop.exp) {
          if (isComponent) {
            dynamicExpr.push(prop.exp)
            pushMergeArg()
            dynamicArgs.push({
              kind: IRDynamicPropsKind.EXPRESSION,
              value: prop.exp,
              handler: true,
            })
          } else {
            context.registerEffect(
              [prop.exp],

              {
                type: IRNodeTypes.SET_DYNAMIC_EVENTS,
                element: context.reference(),
                event: prop.exp,
              },
            )
          }
        } else {
          context.options.onError(
            createCompilerError(ErrorCodes.X_V_ON_NO_EXPRESSION, prop.loc),
          )
        }
        continue
      }
    }

    // exclude `is` prop for <component>
    if (
      (isDynamicComponent &&
        prop.type === NodeTypes.ATTRIBUTE &&
        prop.name === 'is') ||
      (prop.type === NodeTypes.DIRECTIVE &&
        prop.name === 'bind' &&
        isStaticArgOf(prop.arg, 'is'))
    ) {
      continue
    }

    const result = transformProp(prop, node, context)
    if (result) {
      dynamicExpr.push(result.key, result.value)
      if (isComponent && !result.key.isStatic) {
        // v-bind:[name]="value" or v-on:[name]="value"
        pushMergeArg()
        dynamicArgs.push(
          extend(resolveDirectiveResult(result), {
            kind: IRDynamicPropsKind.ATTRIBUTE,
          }) as IRPropsDynamicAttribute,
        )
      } else {
        // other static props
        results.push(result)
      }
    }
  }

  // has dynamic key or v-bind="{}"
  if (dynamicArgs.length || results.some(({ key }) => !key.isStatic)) {
    // take rest of props as dynamic props
    pushMergeArg()
    return [true, dynamicArgs, dynamicExpr]
  }

  const irProps = dedupeProperties(results)
  return [false, irProps]
}

function transformProp(
  prop: VaporDirectiveNode | AttributeNode,
  node: ElementNode,
  context: TransformContext<ElementNode>,
): DirectiveTransformResult | void {
  let { name } = prop

  if (prop.type === NodeTypes.ATTRIBUTE) {
    if (isReservedProp(name)) return
    return {
      key: createSimpleExpression(prop.name, true, prop.nameLoc),
      value: prop.value
        ? createSimpleExpression(prop.value.content, true, prop.value.loc)
        : EMPTY_EXPRESSION,
    }
  }

  const directiveTransform = context.options.directiveTransforms[name]
  if (directiveTransform) {
    return directiveTransform(prop, node, context)
  }

  if (!isBuiltInDirective(name)) {
    const fromSetup = resolveSetupReference(`v-${name}`, context)
    if (fromSetup) {
      name = fromSetup
    } else {
      context.directive.add(name)
    }

    context.registerOperation({
      type: IRNodeTypes.DIRECTIVE,
      element: context.reference(),
      dir: prop,
      name,
      asset: !fromSetup,
    })
  }
}

// Dedupe props in an object literal.
// Literal duplicated attributes would have been warned during the parse phase,
// however, it's possible to encounter duplicated `onXXX` handlers with different
// modifiers. We also need to merge static and dynamic class / style attributes.
function dedupeProperties(results: DirectiveTransformResult[]): IRProp[] {
  const knownProps: Map<string, IRProp> = new Map()
  const deduped: IRProp[] = []

  for (const result of results) {
    const prop = resolveDirectiveResult(result)
    // dynamic keys are always allowed
    if (!prop.key.isStatic) {
      deduped.push(prop)
      continue
    }
    const name = prop.key.content
    const existing = knownProps.get(name)
    if (existing) {
      if (name === 'style' || name === 'class') {
        mergePropValues(existing, prop)
      }
      // unexpected duplicate, should have emitted error during parse
    } else {
      knownProps.set(name, prop)
      deduped.push(prop)
    }
  }
  return deduped
}

function resolveDirectiveResult(prop: DirectiveTransformResult): IRProp {
  return extend({}, prop, {
    value: undefined,
    values: [prop.value],
  })
}

function mergePropValues(existing: IRProp, incoming: IRProp) {
  const newValues = incoming.values
  existing.values.push(...newValues)
}

function isComponentTag(tag: string) {
  return tag === 'component' || tag === 'Component'
}
