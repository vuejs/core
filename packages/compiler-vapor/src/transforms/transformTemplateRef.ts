import {
  NodeTypes,
  type SimpleExpressionNode,
  createSimpleExpression,
} from '@vue/compiler-dom'
import type { NodeTransform } from '../transform'
import { IRNodeTypes } from '../ir'
import { normalizeBindShorthand } from './vBind'
import { findProp, isConstantExpression } from '../utils'
import { EMPTY_EXPRESSION } from './utils'

export const transformTemplateRef: NodeTransform = (node, context) => {
  if (node.type !== NodeTypes.ELEMENT) return

  const dir = findProp(node, 'ref', false, true)
  if (!dir) return

  context.ir.hasTemplateRef = true

  let value: SimpleExpressionNode
  if (dir.type === NodeTypes.DIRECTIVE) {
    value = dir.exp || normalizeBindShorthand(dir.arg!, context)
  } else {
    value = dir.value
      ? createSimpleExpression(dir.value.content, true, dir.value.loc)
      : EMPTY_EXPRESSION
  }

  return () => {
    const id = context.reference()
    const effect = !isConstantExpression(value)
    effect &&
      context.registerOperation({
        type: IRNodeTypes.DECLARE_OLD_REF,
        id,
      })
    context.registerEffect([value], {
      type: IRNodeTypes.SET_TEMPLATE_REF,
      element: id,
      value,
      refFor: !!context.inVFor,
      effect,
    })
  }
}
