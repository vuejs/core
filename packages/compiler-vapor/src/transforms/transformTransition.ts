import type { NodeTransform } from '@vue/compiler-vapor'
import { findDir, isTransitionTag } from '../utils'
import {
  type ElementNode,
  ElementTypes,
  NodeTypes,
  isTemplateNode,
  postTransformTransition,
} from '@vue/compiler-dom'

export const transformTransition: NodeTransform = (node, context) => {
  if (
    node.type === NodeTypes.ELEMENT &&
    node.tagType === ElementTypes.COMPONENT
  ) {
    if (isTransitionTag(node.tag)) {
      return postTransformTransition(
        node,
        context.options.onError,
        hasMultipleChildren,
      )
    }
  }
}

function hasMultipleChildren(node: ElementNode): boolean {
  const children = (node.children = node.children.filter(
    c =>
      c.type !== NodeTypes.COMMENT &&
      !(c.type === NodeTypes.TEXT && !c.content.trim()),
  ))

  const first = children[0]

  // has v-for
  if (
    children.length === 1 &&
    first.type === NodeTypes.ELEMENT &&
    (findDir(first, 'for') || isTemplateNode(first))
  ) {
    return true
  }

  const hasElse = (node: ElementNode) =>
    findDir(node, 'else-if') || findDir(node, 'else', true)

  // has v-if/v-else-if/v-else
  if (
    children.every(
      (c, index) =>
        c.type === NodeTypes.ELEMENT &&
        // not template
        !isTemplateNode(c) &&
        // not has v-for
        !findDir(c, 'for') &&
        // if the first child has v-if, the rest should also have v-else-if/v-else
        (index === 0 ? findDir(c, 'if') : hasElse(c)) &&
        !hasMultipleChildren(c),
    )
  ) {
    return false
  }

  return children.length > 1
}
