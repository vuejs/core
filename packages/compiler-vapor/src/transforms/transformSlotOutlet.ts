import {
  type AttributeNode,
  type ElementNode,
  ElementTypes,
  ErrorCodes,
  NodeTypes,
  type SimpleExpressionNode,
  createCompilerError,
  createSimpleExpression,
  isStaticArgOf,
  isStaticExp,
} from '@vue/compiler-dom'
import type { NodeTransform, TransformContext } from '../transform'
import {
  type BlockIRNode,
  type DirectiveIRNode,
  DynamicFlag,
  IRNodeTypes,
  type IRProps,
  type VaporDirectiveNode,
} from '../ir'
import { camelize, extend } from '@vue/shared'
import { newBlock } from './utils'
import { buildProps } from './transformElement'

export const transformSlotOutlet: NodeTransform = (node, context) => {
  if (node.type !== NodeTypes.ELEMENT || node.tag !== 'slot') {
    return
  }
  const id = context.reference()
  context.dynamic.flags |= DynamicFlag.INSERT | DynamicFlag.NON_TEMPLATE
  const [fallback, exitBlock] = createFallback(
    node,
    context as TransformContext<ElementNode>,
  )

  let slotName: SimpleExpressionNode | undefined
  const slotProps: (AttributeNode | VaporDirectiveNode)[] = []
  for (const prop of node.props as (AttributeNode | VaporDirectiveNode)[]) {
    if (prop.type === NodeTypes.ATTRIBUTE) {
      if (prop.value) {
        if (prop.name === 'name') {
          slotName = createSimpleExpression(prop.value.content, true, prop.loc)
        } else {
          slotProps.push(extend({}, prop, { name: camelize(prop.name) }))
        }
      }
    } else if (prop.name === 'bind' && isStaticArgOf(prop.arg, 'name')) {
      if (prop.exp) {
        slotName = prop.exp!
      } else {
        // v-bind shorthand syntax
        slotName = createSimpleExpression(
          camelize(prop.arg!.content),
          false,
          prop.arg!.loc,
        )
        slotName.ast = null
      }
    } else {
      let slotProp = prop
      if (
        slotProp.name === 'bind' &&
        slotProp.arg &&
        isStaticExp(slotProp.arg)
      ) {
        slotProp = extend({}, prop, {
          arg: extend({}, slotProp.arg, {
            content: camelize(slotProp.arg!.content),
          }),
        })
      }
      slotProps.push(slotProp)
    }
  }

  slotName ||= createSimpleExpression('default', true)
  let irProps: IRProps[] = []
  if (slotProps.length) {
    const [isDynamic, props] = buildProps(
      extend({}, node, { props: slotProps }),
      context as TransformContext<ElementNode>,
      true,
    )
    irProps = isDynamic ? props : [props]

    const runtimeDirective = context.block.operation.find(
      (oper): oper is DirectiveIRNode =>
        oper.type === IRNodeTypes.DIRECTIVE && oper.element === id,
    )
    if (runtimeDirective) {
      context.options.onError(
        createCompilerError(
          ErrorCodes.X_V_SLOT_UNEXPECTED_DIRECTIVE_ON_SLOT_OUTLET,
          runtimeDirective.dir.loc,
        ),
      )
    }
  }

  return () => {
    exitBlock && exitBlock()
    context.dynamic.operation = {
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      id,
      name: slotName,
      props: irProps,
      fallback,
    }
  }
}

function createFallback(
  node: ElementNode,
  context: TransformContext<ElementNode>,
): [block?: BlockIRNode, exit?: () => void] {
  if (!node.children.length) {
    return []
  }

  context.node = node = extend({}, node, {
    type: NodeTypes.ELEMENT,
    tag: 'template',
    props: [],
    tagType: ElementTypes.TEMPLATE,
    children: node.children,
  })

  const fallback = newBlock(node)
  const exitBlock = context.enterBlock(fallback)
  context.reference()
  return [fallback, exitBlock]
}
