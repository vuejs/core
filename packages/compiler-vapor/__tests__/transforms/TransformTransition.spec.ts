import { makeCompile } from './_utils'
import {
  transformChildren,
  transformElement,
  transformText,
  transformVBind,
  transformVIf,
  transformVShow,
  transformVSlot,
} from '@vue/compiler-vapor'
import { transformTransition } from '../../src/transforms/transformTransition'
import { DOMErrorCodes } from '@vue/compiler-dom'

const compileWithElementTransform = makeCompile({
  nodeTransforms: [
    transformText,
    transformVIf,
    transformElement,
    transformVSlot,
    transformChildren,
    transformTransition,
  ],
  directiveTransforms: {
    bind: transformVBind,
    show: transformVShow,
  },
})

describe('compiler: transition', () => {
  test('basic', () => {
    const { code } = compileWithElementTransform(
      `<Transition><h1 v-show="show">foo</h1></Transition>`,
    )
    expect(code).toMatchSnapshot()
  })

  test('v-show + appear', () => {
    const { code } = compileWithElementTransform(
      `<Transition appear><h1 v-show="show">foo</h1></Transition>`,
    )
    expect(code).toMatchSnapshot()
  })

  test('work with v-if', () => {
    const { code } = compileWithElementTransform(
      `<Transition><h1 v-if="show">foo</h1></Transition>`,
    )

    expect(code).toMatchSnapshot()
    // n2 should have a key
    expect(code).contains('n2.$key = 2')
  })

  test('work with dynamic keyed children', () => {
    const { code } = compileWithElementTransform(
      `<Transition>
        <h1 :key="key">foo</h1>
      </Transition>`,
    )

    expect(code).toMatchSnapshot()
    expect(code).contains('_createKeyedFragment(() => _ctx.key')
    // should preserve key
    expect(code).contains('n0.$key = _ctx.key')
  })

  function checkWarning(template: string, shouldWarn = true) {
    const onError = vi.fn()
    compileWithElementTransform(template, { onError })
    if (shouldWarn) {
      expect(onError).toHaveBeenCalled()
      expect(onError.mock.calls).toMatchObject([
        [{ code: DOMErrorCodes.X_TRANSITION_INVALID_CHILDREN }],
      ])
    } else {
      expect(onError).not.toHaveBeenCalled()
    }
  }

  test('warns if multiple children', () => {
    checkWarning(
      `<Transition>
        <h1>foo</h1>
        <h2>bar</h2>
      </Transition>`,
      true,
    )
  })

  test('warns with v-for', () => {
    checkWarning(
      `
      <transition>
        <div v-for="i in items">hey</div>
      </transition>
      `,
      true,
    )
  })

  test('warns with multiple v-if + v-for', () => {
    checkWarning(
      `
      <transition>
        <div v-if="a" v-for="i in items">hey</div>
        <div v-else v-for="i in items">hey</div>
      </transition>
      `,
      true,
    )
  })

  test('warns with template v-if', () => {
    checkWarning(
      `
      <transition>
        <template v-if="ok"></template>
      </transition>
      `,
      true,
    )
  })

  test('warns with multiple templates', () => {
    checkWarning(
      `
      <transition>
        <template v-if="a"></template>
        <template v-else></template>
      </transition>
      `,
      true,
    )
  })

  test('warns if multiple children with v-if', () => {
    checkWarning(
      `
      <transition>
        <div v-if="one">hey</div>
        <div v-if="other">hey</div>
      </transition>
      `,
      true,
    )
  })

  test('does not warn with regular element', () => {
    checkWarning(
      `
      <transition>
        <div>hey</div>
      </transition>
      `,
      false,
    )
  })

  test('does not warn with one single v-if', () => {
    checkWarning(
      `
      <transition>
        <div v-if="a">hey</div>
      </transition>
      `,
      false,
    )
  })

  test('does not warn with v-if v-else-if v-else', () => {
    checkWarning(
      `
      <transition>
        <div v-if="a">hey</div>
        <div v-else-if="b">hey</div>
        <div v-else>hey</div>
      </transition>
      `,
      false,
    )
  })

  test('does not warn with v-if v-else', () => {
    checkWarning(
      `
      <transition>
        <div v-if="a">hey</div>
        <div v-else>hey</div>
      </transition>
      `,
      false,
    )
  })

  test('inject persisted when child has v-show', () => {
    expect(
      compileWithElementTransform(`
        <Transition>
          <div v-show="ok" />
        </Transition>
    `).code,
    ).toMatchSnapshot()
  })

  test('the v-if/else-if/else branches in Transition should ignore comments', () => {
    expect(
      compileWithElementTransform(`
    <transition>
      <div v-if="a">hey</div>
      <!-- this should be ignored -->
      <div v-else-if="b">hey</div>
      <!-- this should be ignored -->
      <div v-else>
        <p v-if="c"/>
        <!-- this should not be ignored -->
        <p v-else/>
      </div>
    </transition>
    `).code,
    ).toMatchSnapshot()
  })
})
