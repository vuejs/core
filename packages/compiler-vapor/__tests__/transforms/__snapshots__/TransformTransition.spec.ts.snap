// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: transition > basic 1`] = `
"import { VaporTransition as _VaporTransition, applyVShow as _applyVShow, createComponent as _createComponent, template as _template } from 'vue';
const t0 = _template("<h1>foo</h1>")

export function render(_ctx) {
  const n1 = _createComponent(_VaporTransition, { persisted: () => ("") }, {
    "default": () => {
      const n0 = t0()
      _applyVShow(n0, () => (_ctx.show))
      return n0
    }
  }, true)
  return n1
}"
`;

exports[`compiler: transition > inject persisted when child has v-show 1`] = `
"import { VaporTransition as _VaporTransition, applyVShow as _applyVShow, createComponent as _createComponent, template as _template } from 'vue';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n1 = _createComponent(_VaporTransition, { persisted: () => ("") }, {
    "default": () => {
      const n0 = t0()
      _applyVShow(n0, () => (_ctx.ok))
      return n0
    }
  }, true)
  return n1
}"
`;

exports[`compiler: transition > the v-if/else-if/else branches in Transition should ignore comments 1`] = `
"import { VaporTransition as _VaporTransition, setInsertionState as _setInsertionState, createIf as _createIf, createComponent as _createComponent, template as _template } from 'vue';
const t0 = _template("<div>hey</div>")
const t1 = _template("<p></p>")
const t2 = _template("<div></div>")

export function render(_ctx) {
  const n16 = _createComponent(_VaporTransition, null, {
    "default": () => {
      const n0 = _createIf(() => (_ctx.a), () => {
        const n2 = t0()
        n2.$key = 2
        return n2
      }, () => _createIf(() => (_ctx.b), () => {
        const n5 = t0()
        n5.$key = 5
        return n5
      }, () => {
        const n14 = t2()
        _setInsertionState(n14, 0)
        const n9 = _createIf(() => (_ctx.c), () => {
          const n11 = t1()
          return n11
        }, () => {
          const n13 = t1()
          return n13
        })
        n14.$key = 14
        return n14
      }))
      return [n0, n3, n7]
    }
  }, true)
  return n16
}"
`;

exports[`compiler: transition > v-show + appear 1`] = `
"import { VaporTransition as _VaporTransition, applyVShow as _applyVShow, createComponent as _createComponent, template as _template } from 'vue';
const t0 = _template("<h1>foo</h1>")

export function render(_ctx) {
  const deferredApplyVShows = []
  const n1 = _createComponent(_VaporTransition, {
    appear: () => (""), 
    persisted: () => ("")
  }, {
    "default": () => {
      const n0 = t0()
      deferredApplyVShows.push(() => _applyVShow(n0, () => (_ctx.show)))
      return n0
    }
  }, true)
  deferredApplyVShows.forEach(fn => fn())
  return n1
}"
`;

exports[`compiler: transition > work with dynamic keyed children 1`] = `
"import { VaporTransition as _VaporTransition, createKeyedFragment as _createKeyedFragment, createComponent as _createComponent, template as _template } from 'vue';
const t0 = _template("<h1>foo</h1>")

export function render(_ctx) {
  const n1 = _createComponent(_VaporTransition, null, {
    "default": () => {
      return _createKeyedFragment(() => _ctx.key, () => {
        const n0 = t0()
        n0.$key = _ctx.key
        return n0
      })
    }
  }, true)
  return n1
}"
`;

exports[`compiler: transition > work with v-if 1`] = `
"import { VaporTransition as _VaporTransition, createIf as _createIf, createComponent as _createComponent, template as _template } from 'vue';
const t0 = _template("<h1>foo</h1>")

export function render(_ctx) {
  const n3 = _createComponent(_VaporTransition, null, {
    "default": () => {
      const n0 = _createIf(() => (_ctx.show), () => {
        const n2 = t0()
        n2.$key = 2
        return n2
      })
      return n0
    }
  }, true)
  return n3
}"
`;
